lib/utils/buildHelper.js

```diff
@@ -0,0 +1,128 @@
+ *
+ * @param {object} parameters Parameters
+ * @param {object} parameters.tree Project tree as generated by the
+ *   [@ui5/project.normalizer]{@link module:@ui5/project.normalizer}
+ * @param {Array.<string>} parameters.dependencies The dependencies to be considered
+ * @param {Array.<string>} [parameters.dependenciesRegExp] The dependencies to be considered as new
+ *   RegExp instances
+ * @param {boolean} [parameters.handleSubtree=false] Decides whether the sub-dependencies of the given
+ *   dependencies are included into the returned dependency list
+ * @returns {Array.<string|RegExp>} Array of the computed dependencies
+ */
+function createDependencyList({tree, dependencies = [], dependenciesRegExp = [], handleSubtree}) {
+	const dependencyList = [];
+	if (handleSubtree && dependenciesRegExp.length) {
+		throw new Error("RegExp's should not be appended to list of sub-dependencies");
+	}
+
+	function _addUnique(depName) {
+		if (!dependencyList.includes(depName)) {
+			dependencyList.push(depName);
+		}
+	}
+	function _findDependency(searchProject, depName) {
+		let project = searchProject.dependencies.find((dep) => dep.metadata.name === depName);
+
+		if (!project) {
+			for (const project0 of searchProject.dependencies) {
+				const result = _findDependency(project0, depName);
+				if (result) {
+					project = result;
+					break;
+				}
+			}
+		}
+
+		return project;
+	}
+	function _processSubtree(project) {
+		project.dependencies.forEach((project0) => {
+			_addUnique(project0.metadata.name);
+			_processSubtree(project0);
+		});
+	}
+
+	dependencies.forEach((depName) => {
+		if (depName === "*") {
+			return _addUnique(depName);
+		}
+		const project = _findDependency(tree, depName);
+		if (project) {
+			_addUnique(depName);
+			if (handleSubtree) {
+				_processSubtree(project);
+			}
+		} else {
+			log.warn(`Could not find dependency '${depName}' for '${tree.metadata.name}'.`);
+		}
+	});
+
+	return dependencyList.concat(dependenciesRegExp.map((exp) => new RegExp(exp)));
+}
+
+/**
+ * Merges a <code>dependencyList</code> into a <code>targetDependencyList</code> without duplicates. Values
+ * given in <code>excludeList</code> are skipped.
+ *
+ * @param {Array.<string|RegExp>} targetDependencyList The target list
+ * @param {Array.<string|RegExp>} dependencyList The source list
+ * @param {Array.<string|RegExp>} excludeList The list of values to be excluded from the merge operation
+ */
+function mergeDependencyLists(targetDependencyList, dependencyList, excludeList) {
+	function isInList(depList, depName) {
+		return depList.some((dep) => dep instanceof RegExp ? dep.test(depName) : dep === depName);
+	}
+	function isRegExpInList(depList, depRegExp) {
+		return depList.some((dep) => dep instanceof RegExp && dep.toString() === depRegExp.toString());
+	}
+	dependencyList.forEach((dep) => {
+		if (
+			dep instanceof RegExp && !isRegExpInList(targetDependencyList, dep) && !isRegExpInList(excludeList, dep) ||
```

Comment:
the check `!isRegExpInList(excludeList, dep))` looks tricky to me. I understand that it is balanced with the `!isInList(...)`check in the next line. But while the `!isInList(...)`check gives an exact result, this one here is only an approximation. There will be cases where an includeRegExp (/^ab/) is totally overruled by an excludeRegExp (e.g. `/^a/`), but the check can't detect this.

I guess the answer will will also answer my question above, why the builder still requires include and exclude options and why a preprocessing of the tree is not enough. There must be a later phase that handles the regexp correctly. But if so, why not totally leave it to that phase?

Or the other way around: why is it not possible to calculate the final list of "projects to be built" and give that to the builder? Looks strange to me that the filtering has to be split across layers.

Reply:
We use this function on three occasions:

1. Merging the resolved include-trees (which do not contain regular expressions) into the includes list
2. Merging the resolved exclude-trees (which do not contain regular expressions) into the excludes list
3. Merging the includes (string, regex and resolved trees) from the ui5.yaml configuration into the includes list

I think if we find a different solution for merging 3 into the list of includes, we can simplify this function quite a bit, since `dependencyList` can then never be a regular expression.

As for merging the ui5.yaml defaults into the list, we could collet the regular expression includes in a separate variable which is always pushed onto the list and only use mergeDependencyLists for the "string" includes and the trees.

Reply:
As described in https://github.com/SAP/ui5-cli/pull/442#discussion_r671059169
This priority issue and comparison is improved now. RegExps are no longer considered separately.